### ISSUE TITLE
Option to lock the scale between multiple images

### ISSUE THREAD

**Issue by o-smirnov** [ericmandel/js9#45]
I've got the partners code working with multiple images now. A very useful use case is to lock the same color/scale across multiple images, then to compare them using blink (or the next/prev image command). Would this be a difficult thing to add to the "Scale" menu? I can also implement an external widget for it in the partners code -- I have a pretty clear idea of how to do it now -- but that would probably look a little ugly...

**Comment by ericmandel** [comment]
Are you perhaps thinking about a new public routine like this:
````
# 1st arg specifies the type of change that gets sync'ed between images
# 2nd arg not specified (all images), one image handle/id, or array of handles/ids, or an return sid
sid = JS9.Sync(["colormap", "pan", "scale", "wcssys", "wcsunits", "zoom"])   # sync all images
sid = JS9.Sync(["colormap", "pan", "scale", "wcssys", "wcsunits", "zoom"], [im1, im2]) # sync specified images
sid = JS9.Sync(["colormap", ...], sid)    # change types that get synced for existing sync setup
sid = JS9.sync([] or null, sid)                # remove existing sync setup
````
So you specify one or more sync types and zero or more images or returned id, and onset[type] callbacks ensure that changing one image causes the same to happen on the other images?

That sounds like a pretty good idea. But perhaps you mean something else?

**Comment by o-smirnov** [comment]
Yeah, a ``JS9.Sync()`` routine like that would be excellent. Currently, I have a lot of fiddly code in the partner display implementation which is trying to achieve the same thing, but if that was a native JS9 capability, things would look a lot cleaner.

Some open questions then:

* What does it mean to pan-sync two images of different sizes? If you pan to a region of the bigger image that is outside the smaller image, what do you then show for the smaller image? Can you tell JS9 to pan to a coordinate outside the image (which would make sense I guess -- you'd still want to keep the visible sections aligned...)?

* Would you support user controls for this in the JS9 menus? For my application, this is not strictly necessary -- I've figured out how to insert widgets next to the menubar:

![image](https://user-images.githubusercontent.com/6470079/44302479-e5408e80-a329-11e8-8d00-cbffab04dd5f.png)

...so I can use the "lock" button to enable/disable scale-sync (pan-zoom-sync is something I want to be always on anyway). 

* The widgets actually belong (from a UX point of view) next to the colorbar, but I couldn't achieve the same there due to some unholy CSS interaction. It would be nice to put a "<->" button next to it that would activate the "Scale | set limits to data min/max" function.

**Comment by o-smirnov** [comment]
> I probably need to know more about why you could not just add stuff in between the buttons

Probably just out of ignorance... are you suggesting I call ``.append()`` in Javascript on the ``JS9Menus_JS9Menubar`` element to get my custom buttons in place? I could do that, sure. Although at this point it feels like I'm relying on the structure of the internals, rather than on the public API, so I'd be a little hesitant about doing that, since I assume it may be subject to change.

**Comment by o-smirnov** [comment]
> BTW, have you seen this: https://js9.si.edu/js9/js9menustyles.html and the underlying user definition file for user menus: https://js9.si.edu/js9/js9usermenus.js. It might be of some interest ...

I have not, but that's an excellent pointer, since my next question was about customizing menus...I'll get around to that at some point!

**Comment by ericmandel** [comment]
> are you suggesting I call .append() in Javascript on the JS9Menus_JS9Menubar element to get my custom buttons in place? Although at this point it feels like I'm relying on the structure of the internals, rather than on the public API, 

Yeah ... this is a problem that I don't know how to handle. I mean, I can't really parameterize everything, but that does leave you in the uncomfortable position of going outside the API.

I'm not sure what to say, and would welcome your opinion about how to handle things better.

**Comment by o-smirnov** [comment]
> But perhaps you are thinking of direct menu access to more primitive "sync the colormap between all images" and "sync the scale between all images" options, in the Colormap and Scale menus?

That's exactly what I'm thinking about...

> Yeah ... this is a problem that I don't know how to handle. I mean, I can't really parameterize everything, but that does leave you in the uncomfortable position of going outside the API.

That's all right -- maybe if you just gradually promote these things to the documentation ("the DIV elements are named as follows..."), then it become part of the API effectively, and disincentivizes you from renaming them in the future...

**Comment by ericmandel** [comment]
> That's exactly what I'm thinking about...

OK, got it, I just need to finish a scaling plugin while I mull over JS9.Sync() ...

> maybe if you just gradually promote these things to the documentation ("the DIV elements are named as follows..."), then it become part of the API effectively, and disincentivizes you from renaming them in the future...

Good idea and  thanks for using the word "gradually". But also ha! when I looked at the actual button id's I thought "hmmm ... there should be a hyphen between the menu name and the display class id,   and shouldn't the display class come first?" ... so there will have to be some changes to make the id's consistent if they are going to be documented (though I try to maintain backward compatibilitywhen I do make consistency changes).

**Comment by ericmandel** [comment]
Experiment 2 with the sync plugin is here: 

https://gist.github.com/ericmandel/0084f652cefe3269067bde1a79a16818

There are a number of vexing problems:
-- I haven't looked at the panning issues yet, probably will require changes to SetPan itself.
-- It's not clear how to sync regions, I might have to tag them specially (but not export the sync tags), but what to do with already-existing regions?
-- not clear how to sync regions, I might have to save the changes made to a region in one display, so they can be applied to a region in another display ... uggh
-- given the number of options, what sort of UI (other than a complicated one) would be of any use at all? Seems like a general "just sync everyone to everything" would not be good (and might be very annoying) in lots of situations.

Once I did see an enormous slowdown in redrawing the images as contrast/bias changed, which you mentioned e: partner sync'ing. This looks like the performance you get if the GPU is not being used (circa 2012, it happened a lot while browsers got their GPU support worked out). Not sure what we can do about that ... but it might be worthwhile noticing what else is happening on the machine when this takes place.

It's Labor Day here in the US, and its almost a sin to work today, so responses will be sporadic.

**Comment by o-smirnov** [comment]
Thanks! I'm about to hit the road/take leave/etc., so it'll take me a while to absorb these changes. So I have left you to your long weekend in peace...

But in a couple of weeks or so, I should be bothering you again.

> -- It's not clear how to sync regions

I agree, that sounds like a can of worms. I don't *really* need it if you don't?

> -- given the number of options, what sort of UI (other than a complicated one) would be of any use at all? 

What I've found most useful is locking the scales and colormaps of all images (or a subset of images) together. So as the simplest UI control, that would make sense.

Not that I expect to use the UI component in the partners code anyway -- there I need something more tricky still -- i.e. each rebinned and zoomed pair of images must always be locked to each other, but multiple pairs of such images may or may not be locked together... On the one hand, I've got it implemented programmatically using an external widget. On the other hand, I don't see any reasonably simple way to expose this as an option in a JS9 menu, so I'm disinclined to bother you about it.

**Comment by ericmandel** [comment]
All good info, thanks, I'll keep thinking about this.

And I need to talk to some people here about how they might use this functionality ... once they return from IAU2018. A more immediate concern will be with panning in general, since JS9 currently (and deliberately) protects you from panning off the image. That constraint needs to be relaxed in general, or via parameter-driven configuration?

Have a good and safe trip!

**Comment by ericmandel** [comment]
In case you are back from traveling ... I have implemented image sync, a demo is at: https://js9.si.edu/js9/js9sync.html

The File menu has sync and unsync options, but you'll probably just use the JS9.SyncImages() routine to set up the partners.

It turns out that region sync'ing is going to be very useful in our X-ray neck of the woods ... and also was the hardest part to implement.

Caveat: I had to punt on panning beyond the image boundary when one image is larger than the other: the smaller image still stops panning at the edge. Turns out a of code breaks when I tried to relax that condition, so we'll leave that for another time. (This might be the preferable solution in any case.)

**Comment by o-smirnov** [comment]
Excellent, thanks! I got back from one trip, only to find myself neck-deep in a complete office relocation... so all real work has been on the back burner. I do hope to spend some time next week setting up a pipeline visualization system based on what we've done so far, so you'll surely be hearing from me then.

**Comment by ericmandel** [comment]
Welcome back! Development work (and bug reports) having quieted down, I plan to release v2.2 in the next week or so ... depending on what my XMM-Newton colleagues uncover in the meantime. But we can pick things up via continuous improvements in GitHub, as usual.
